File: 95 Unique Binary Search Tree ii/unique_bst_2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        def rec(start, end):
            if start > end:
                return [None]

            if start == end:
                return [TreeNode(start)]

            ret_list = []

            for i in range(start, end+1):
                left = rec(start, i-1)
                right = rec(i+1, end)

                for pair in product(left, right):
                    ret_list.append(TreeNode(i, pair[0], pair[1]))

            return ret_list

        res = rec(1, n)
        return res

File: 95 Unique Binary Search Tree ii/unique_bst_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        return rec(1,n);
    }
    
    public List<TreeNode> rec(int start, int end) {
        List<TreeNode> res = new ArrayList();
        if ( start > end ) {
            res.add(null);
            return res;
        }
        for(int i=start; i<=end; i++) {
            List<TreeNode> left = rec(start,i-1);
            List<TreeNode> right = rec(i+1,end);
            
            for(TreeNode l : left ) {
                for(TreeNode r : right ) {
                    TreeNode current = new TreeNode(i);
                    current.left = l;
                    current.right = r;
                    res.add(current);
                }
            }
        }
        return res;
    }
}